addons/purchase/tests/test_purchase_downpayment.py

```diff
@@ -0,0 +1,686 @@
+
+        downpayment_line = self.purchase_order.order_line.filtered(lambda l: l.is_downpayment and not l.display_type)
+        self.assertEqual(downpayment_line[0].price_unit, 50, 'The down payment unit price should not change on PO')
+        self.post_bills()
+        self.assertEqual(downpayment_line[0].price_unit, 50, 'The down payment unit price should not change on PO')
+
+    def test_downpayment_fixed_amount_with_zero_total_amount(self):
+        """
+            Check that DP should be zero if the total PO amount is zero.
+        """
+        # PO with one line and amount total is zero
+        self.pol_serv_deliver.unlink()
+        self.pol_serv_order.unlink()
+        self.pol_product_deliver.unlink()
+        self.pol_product_order.product_qty = 5
+        self.pol_product_order.price_unit = 0
+
+        self.purchase_order.button_confirm()
+        self.purchase_order.order_line.qty_received = 5
+        self.make_downpayment(advance_payment_method='fixed', fixed_amount=50)
+        self.assertEqual(self.purchase_order.order_line[2].price_unit, 0.0, 'The down payment amount should be 0.0')
+
+    def test_downpayment_percentage_tax_icl(self):
+        """
+            Test invoice with a percentage downpayment and an included tax.
+            Check that the total amount of the invoice is correct and equal to the PO total amount.
+        """
+        self.purchase_order.button_confirm()
+        self.make_downpayment()
+
+        self.assertEqual(len(self.purchase_order.invoice_ids), 1, 'Invoice should be created for the PO')
+        downpayment_line = self.purchase_order.order_line.filtered(lambda l: l.is_downpayment and not l.display_type)
+        self.assertEqual(len(downpayment_line), 1, 'PO line downpayment should be created on PO')
+        self.assertEqual(downpayment_line.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+
+        invoice = self.purchase_order.invoice_ids[0]
+        downpayment_aml = invoice.line_ids.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))[0]
+        self.assertEqual(downpayment_aml.price_total, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+        self.assertEqual(downpayment_aml.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+        self.post_bills()
+        self.assertEqual(downpayment_line.price_unit, self.purchase_order.amount_total / 2, 'downpayment should have the correct amount')
+
+    def test_downpayment_invoice_and_partial_credit_note(self):
+        """This test check that the downpayment line amount on the purchase order remains consistent"""
+        self.purchase_order.button_confirm()
+
+        self.make_downpayment(advance_payment_method='fixed', fixed_amount=100)
+
+        # Ensure the downpayment line on the sale order is correctly set to 100
```

Comment:
`sale order => purchase order`

I guess that it's supposed to be a refund in the comment + assert message below as well